### PR TITLE
[filebeat][filestream] Enable status reporter for filestream input

### PR DESCRIPTION
**Pull request by VihasMakwana**
## Proposed commit message

Enable `StatusReporter` for `filestream` input which was introduced in https://github.com/elastic/beats/pull/39209

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Closes [39734](https://github.com/elastic/beats/issues/39734)
## Use cases

This PR allows the filestream to report its status to elastic-agent. This would keep the status of the unit up-to-date and would help with diagnostics.

### CODE REVIEW

File: CHANGELOG.next.asciidoc [hunk 1]
```diff
@@ -44,6 +44,7 @@ https://github.com/elastic/beats/compare/v8.8.1\...main[Check the HEAD diff]
 - Update Salesforce module to use new Salesforce input. {pull}37509[37509]
 - Tag events that come from a filestream in "take over" mode. {pull}39828[39828]
 - Fix high IO and handling of a corrupted registry log file. {pull}35893[35893]
+- Enable StatusReporter for filestream input. {pull}40121[40121]
```
Comment (belimawr) [inline CHANGELOG.next.asciidoc:47]: The `CHANGELOG.next.asciidoc` is the public facing changelog that goes into the release notes, we should write entries from the perspective of a user that does not know the internal details of our applications. A user looking to deploy Filebeat does not know what `StatusReporter` means.

You can phrase it by describing the observable behaviour, something like: *"Filebeat, when running under Elastic-Agent" reports detailed status for the Filestream input"*

File: filebeat/input/filestream/internal/input-logfile/harvester.go [hunk 2]
```diff
@@ -204,7 +205,7 @@ func startHarvester(
 			if errors.Is(err, ErrHarvesterAlreadyRunning) {
 				return nil
 			}
-
+			ctx.UpdateStatus(status.Degraded, err.Error())
 			return fmt.Errorf("error while adding new reader to the bookkeeper %w", err)
```
Comment (belimawr) [inline filebeat/input/filestream/internal/input-logfile/harvester.go:209]: By just reporting the message from `err.Error()` we're not sending enough context for the user, it is best to also add the message from the returned error:
```suggestion
			err = fmt.Errorf("error while adding new reader to the bookkeeper %w", err)
			ctx.UpdateStatus(status.Degraded, err.Error())
			return err
```

The same applies to all other places you added a `ctx.UpdateStatus` right above an error return.

**Review by belimawr — CHANGES_REQUESTED** [review]

File: filebeat/input/filestream/internal/input-logfile/harvester.go (hunk 2 above)
Comment (VihasMakwana) [reply filebeat/input/filestream/internal/input-logfile/harvester.go]: Done.

**Review by belimawr — APPROVED** [review]

File: filebeat/input/filestream/input.go [hunk 3]
```diff
@@ -163,7 +164,11 @@ func (inp *filestream) Run(
 	})
 	defer streamCancel()
 
-	return inp.readFromSource(ctx, log, r, fs.newPath, state, publisher, metrics)
+	if err := inp.readFromSource(ctx, log, r, fs.newPath, state, publisher, metrics); err != nil {
+		ctx.UpdateStatus(status.Degraded, fmt.Sprintf("error while reading from source: %v", err))
```
Comment (cmacknz) [inline filebeat/input/filestream/input.go:168]: Sorry for the late comment, if this exits with a transient error, what is setting the input state back to healthy?

In general the thing I am watching out for in the status reporting PRs situations where we permanently mark inputs as degraded or failed after the error condition has cleared.

Comment (VihasMakwana) [reply filebeat/input/filestream/input.go:168]: @cmacknz I understand your concern. 
I did some digging and turns out that `readFromSource` only throws an error when it calls `Publish()`
https://github.com/elastic/beats/blob/d87d15b406b0162989955f3f43053ff77192d415/filebeat/input/filestream/input.go#L394-L397
The above error is return when the context is cancelled. 
https://github.com/elastic/beats/blob/d87d15b406b0162989955f3f43053ff77192d415/filebeat/input/filestream/internal/input-logfile/publish.go#L91-L97.

For other cases (EOF, reader was closed etc.), we just logs the error and return `nil`. So, we will not mark the input as degraded, as they are transient in nature.

Comment (VihasMakwana) [reply filebeat/input/filestream/input.go:168]: @cmacknz maybe we don't need this particular `ctx.UpdateStatus`? I don't think we need to mark input as degraded when the context is cancelled (which should ideally happen during beat shutdown).

I overlooked this particular `ctx.UpdateStatus`.

**Pull request merged**